tests/migrations/test_dataall_migrations.py

```diff
@@ -1,71 +1,190 @@
 
 
-def test_dataall_full_upgrade_success():
-    current_key = TestInitMigration.revision_id()
-    manager = MigrationManager(current_key, TestInitMigration)
+def get_migration(revision_id, next_migration, up=None, down=None):
+    mock_name = f'Migration{revision_id}'
+    migration = MagicMock(name=mock_name, spec=MigrationBase)
+    migration.next_migration.return_value = next_migration
+    migration.revision_id.return_value = revision_id
+    migration.__name__ = mock_name
+    migration.up.side_effect = up
+    migration.down.side_effect = down
+    return migration
+
+
+def test_dont_call_current_migration():
+    m2 = get_migration('2', None)
+    m1 = get_migration('1', m2)
+    m0 = get_migration('0', m1)
+
+    manager = MigrationManager('0', m0)
     new_version = manager.upgrade()
-    assert new_version == Test2Migration.revision_id()
+
+    m0.up.assert_not_called()
+    m1.up.assert_called_once()
+    m2.up.assert_called_once()
+
+    assert_that(new_version).is_equal_to('2')
 
 
 def test_dataall_partial_upgrade_success():
-    current_key = TestInitMigration.revision_id()
-    manager = MigrationManager(current_key, TestInitMigration)
-    new_version = manager.upgrade(TestMigration.revision_id())
-    assert new_version == TestMigration.revision_id()
+    m3 = get_migration('3', None)
+    m2 = get_migration('2', m3)
+    m1 = get_migration('1', m2)
+    m0 = get_migration('0', m1)
 
+    # upgrade to the middle
+    manager = MigrationManager('1', m0)
+    new_version = manager.upgrade('2')
 
-def test_dataall_full_downrade_success():
-    current_key = Test2Migration.revision_id()
-    manager = MigrationManager(current_key, TestInitMigration)
-    new_version = manager.downgrade()
-    assert new_version == TestInitMigration.revision_id()
+    m0.up.assert_not_called()
+    m1.up.assert_not_called()
+    m2.up.assert_called_once()
+    m3.up.assert_not_called()
 
+    assert_that(new_version).is_equal_to('2')
 
-def test_dataall_partial_downrade_success():
-    current_key = Test2Migration.revision_id()
-    manager = MigrationManager(current_key, TestInitMigration)
-    new_version = manager.downgrade(TestMigration.revision_id())
-    assert new_version == TestMigration.revision_id()
+    # continue upgrade to the end with the same manager
+    new_version = manager.upgrade()
+
+    m0.up.assert_not_called()
+    m1.up.assert_not_called()
+    m2.up.assert_called_once()
+    m3.up.assert_called_once()
+
+    assert_that(new_version).is_equal_to('3')
 
 
 def test_incorrect_current_revison_key():
-    current_revision = 'incorrect'
-    try:
-        manager = MigrationManager(current_revision, TestInitMigration)
-    except Exception as e:
-        assert f'{e}' == f'Revision {current_revision} is not found.'
+    incorrect_key = 'incorrect'
+
+    m3 = get_migration('3', None)
+    m2 = get_migration('2', m3)
+    m1 = get_migration('1', m2)
+    m0 = get_migration('0', m1)
```

Comment:
nit: you can DRY it tests by creating 2 fixtures one that returns m0->m3 and one m0->m3 with m2 throwing.